src/React/Basic/Hooks/Internal.purs

```diff
@@ -0,0 +1,127 @@
+module React.Basic.Hooks.Internal
+  ( Render
+  , newtypeHook
+  , unsafeHook
+  , unsafeRenderEffect
+  , Pure
+  , Hook
+  , bind
+  , discard
+  ) where
+
+import Prelude hiding (bind)
+import Control.Applicative.Indexed (class IxApplicative)
+import Control.Apply.Indexed (class IxApply)
+import Control.Bind.Indexed (class IxBind, ibind)
+import Data.Functor.Indexed (class IxFunctor)
+import Data.Newtype (class Newtype)
+import Effect (Effect)
+import Prelude (bind) as Prelude
+import Type.Equality (class TypeEquals)
+
+-- | Render represents the effects allowed within a React component's
+-- | body, i.e. during "render". This includes hooks and ends with
+-- | returning JSX (see `pure`), but does not allow arbitrary side
+-- | effects.
+newtype Render x y a
+  = Render (Effect a)
+
+-- | Alias a chain of hooks. Useful for exposing a single "clean"
+-- | type when creating a hook to improve error messages
+-- | and hide implementation details.
+-- |
+-- | For example, the following alias is technically correct but
+-- | when inspecting types or error messages the alias is expanded
+-- | to the full original type:
+-- |
+-- | ```purs
+-- | type UseNodeDistance hooks = UseEffect Unit (UseState Int (UseRef (Nullable Node) hooks))
+-- |
+-- | useNodeDistanceFromMouse :: Hook UseNodeDistance (Int /\ (Ref (Nullable Node)))
+-- | ```
+-- |
+-- | `aliasHook` allows TODO!
+-- | when inspecting types or error messages the alias is expanded
+-- | to the full original type:
+-- |
+-- | ```purs
+-- | type UseNodeDistance hooks = UseEffect Unit (UseState Int (UseRef (Nullable Node) hooks))
+-- |
+-- | useNodeDistanceFromMouse :: Hook UseNodeDistance (Int /\ (Ref (Nullable Node)))
+-- | ```
+-- |
+-- |
+-- |
+newtypeHook ::
+  forall hooks oldHook newHook a.
+  Newtype newHook oldHook =>
+  Render hooks oldHook a ->
+  Render hooks newHook a
+newtypeHook (Render a) = Render a
+
+-- | Promote an arbitrary Effect to a Hook.
+-- |
+-- | This is unsafe because it allows arbitrary
+-- | effects to be performed during a render, which
+-- | may cause them to be run many times by React.
+-- | This function is primarily for constructing
+-- | new hooks using the FFI. If you just want to
+-- | alias a safe hook's effects, prefer `newtypeHook`.
+unsafeHook ::
+  forall newHook a.
+  Effect a -> Hook newHook a
+unsafeHook = Render
+
+-- | Promote an arbitrary Effect to a Pure render effect.
+-- |
+-- | This is unsafe because it allows arbitrary
+-- | effects to be performed during a render, which
+-- | may cause them to be run many times by React.
+-- | You should almost always prefer `useEffect`!
+unsafeRenderEffect :: forall a. Effect a -> Pure a
+unsafeRenderEffect = Render
```

Comment:
These two unsafe functions are named unsafe because of the effects they might produce if used incorrectly, but I'm not sure if that _actually_ makes them unsafe..